src/components/Download/DatasetDetails/SpeciesView/index.js

```diff
@@ -2,63 +2,62 @@ import unionizeArrays from 'helpers/unionizeArrays'
 import { ViewBlock } from './ViewBlock'
 import { ViewBlocks } from '../ViewBlocks'
 
-export const SpeciesView = ({
-  dataset: {
-    data: datasetData,
-    id: datasetId,
-    experiments,
-    organism_samples: samplesBySpecies,
-    quantile_normalize: quantileNormalize
-  },
-  isImmutable
-}) => {
+export const SpeciesView = ({ dataset, isImmutable }) => {
+  const { organism_samples: organismSamples, experiments } = dataset
+  const organismsNames = Object.keys(organismSamples)
+  // format experiments into an object with accession codes as ksey
+  const formattedExperiments = Object.fromEntries(
+    experiments.map((experiment) => [experiment.accession_code, experiment])
+  )
+
+  // filter the dataset.data to only include the experiments containing given samples
+  const getDataSlice = (samples) => {
+    return Object.entries(dataset.data).reduce((accumulator, [key, value]) => {
+      const filteredSamples = value.filter((accessionCode) =>
+        samples.includes(accessionCode)
+      )
+      if (filteredSamples.length) {
+        accumulator[key] = filteredSamples
+      }
+      return accumulator
+    }, {})
+  }
+
+  // merge all the sample metadata fields from experiments in the organism data slice
+  // to display all possible values of these samples in the samples table
+  const getSampleMetadata = (dataSlice) => {
+    return unionizeArrays(
+      ...Object.keys(dataSlice).map(
+        (accessionCode) =>
+          formattedExperiments[accessionCode]?.sample_metadata || []
+      )
+    )
+  }
```

Comment:
I think when you want to mutate one object into another we can do that up front at little cost. Below is an implementation that extracts two parts of logic to allow for a clear change from mapping organism -> samples to organism -> experiments metadata


```suggestion
// generic helper for determining if there is overlap
const hasIntersect = (arr1, arr2) => arr2.find((item) => arr1.includes(item))

// function that filters the dataset experiments based on passed in samples and returns the metadata for those experiments.
const getSamplesMetadata = (samples) =>
    Object.entries(dataset.data)
        .filter(([, experimentSamples]) => hasIntersect(samples, experimentSamples))
        .map(([accession]) => dataset.experiments.find(e => e.accession === accession).samples_metadata)

// mapping the original object to the new one
const organismMetadata = Object.fromEntries(
    Object.entries(organismSamples).map((organism, samples) => {
        return [organism, getSamplesMetadata(samples)]
    })
)
```

I think this is a good place to start but I am open to feedback. Let me know what you would change. Either way, if we can create the data structure up front once, then we should opt for that. I think maybe it might make sense to split up `getSamplesMetadata` into the filtering and the converting.